prune the heap more

gburd opened the pull request.

gburd committed:
dev setup v12

gburd committed:
Prune heap page on INSERT when the page is considered full

Exercise the existing heapam page pruning code on the INSERT paths to
more pro-actively re-order heap pages and potentially free up space.

gburd committed:
Proactively prune a page when the new tuple could fit.

heap_page_prune_opt() doesn't take into account the size of the tuple that's
trying to be inserted, it only checks if the page has less free space than the
minimum threshold, but doesn't consider whether pruning could make enough space
for a specific tuple.  Consider pruning if the current free space plus
potential space from pruning could accommodate the tuple, even if the page
doesn't meet the normal threshold.

melanieplageman commented on src/backend/access/heap/heapam.c:
```diff
@@ -2243,7 +2243,18 @@ heap_insert(Relation relation, HeapTuple tup, CommandId cid,
+	 * Consider pruning the page if it's getting full, following the same
+	 * heuristics used for UPDATE operations. This helps maintain page density
+	 * and can free up space for future insertions.
+	 */
```
opportunistic pruning happens on update and delete when we sequentially scan to get the page that has the tuples targeted for modification. so it isn't explicitly done in the update code path, thus I don't think it really makes sense to say what you are doing has similar logic to pruning on update.

It actually could make sense to add more explicit opportunistic pruning to update -- for example when the tuple doesn't fit on the modified page and we are looking for a new page and there isn't enough space on that new one we get from the FSM.

pruning explicitly after insert could work. conditioning it on the page being full means it won't be called too often. I'd have to think more about this. I know it was proposed in the past and can't remember what happened to it. They might have been doing it in RelationGetBufferForTuple() before inserting the tuple. Which doesn't seem necessary unless the FSM was wrong about the block having enough space for the inserted tuple.

gburd replied:
Okay, I agree it makes sense not to call out that it is similar to pruning on update, removed.

Right, that's why I proposed pruning when there might be enough space on a page before deciding if HOT is an option.

I'll look for threads that exist on the topic.

melanieplageman replied:
well, we do prune the page when it is being scanned. so, I suppose there is a chance that between then and doing the update to the page that other stuff has happened allowing further HOT pruning (I'd have to think more about if that's possible). Is that what you are imagining?

melanieplageman commented on src/backend/access/heap/heapam.c:
```diff
@@ -569,7 +569,7 @@ heap_prepare_pagescan(TableScanDesc sscan)
 	 * Prune and repair fragmentation for the whole page, if possible.
 	 */
-	heap_page_prune_opt(scan->rs_base.rs_rd, buffer);
+	heap_page_prune_opt(scan->rs_base.rs_rd, buffer, 0);
```
I'm confused. Where do you plan to get this info? I must need to look at other commits.

gburd committed:
Prune when possible during scan of heap pages

Add scan-time heap page pruning functionality and a mechanism to track page
modifications during scans and re-evaluate pruning eligibility.  This addresses
the issue where multiple updates during a scan don't trigger heap page pruning
until the next scan.

gburd committed:
Mark pages prunable during insert operations for abort cleanup

Set PageSetPrunable during heap_insert() and heap_multi_insert() to
enable opportunistic cleanup of dead tuples from aborted transactions.
Previously, aborted insert tuples would remain until the next vacuum
cycle if no other modifications occurred on the page.

gburd committed:
DO NOT COMMIT - TESTING CODE
